### PR TITLE
[FLINK-25178][table] Throw exception when managed table sink needs checkpointing

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*In the past, many users encountered the problem that the sink did not output because they did not open the checkpoint. For the managed table: The planner will throw an exception if the checkpoint is not turned on. (Later we can add public connector interface, including Filesystem, Hive, Iceberg, Hudi need it). `PlannerBase` adds the check whether the managed table sink is enabled to checkpointing.*

## Brief change log

  - *`PlannerBase` adds the `checkManagedTableCheckpointing` to check whether the managed table sink is enabled to checkpointing. If the managed table sink disables checkpointing, this could throw the `TableException`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala [hunk 1]
```diff
@@ -533,4 +536,44 @@ abstract class PlannerBase(
...
+    if (isManagedTable(catalog, table) &&
+      !executor.asInstanceOf[DefaultExecutor]
+        .getExecutionEnvironment
+        .getCheckpointConfig
+        .isCheckpointingEnabled) {
+      throw new TableException(s"Managed sink ${objectIdentifier} should enable the checkpointing.")
+    }
+  }
+
+  private def isManagedTable(catalog: Option[Catalog],
```
Comment (JingsongLi) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala:553]: You can reuse `ManagedTableListener.isManagedTable`, make it static

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala [hunk 2]
```diff
@@ -370,6 +371,7 @@ abstract class PlannerBase(
     lookupResult.getTable match {
       case connectorTable: ConnectorCatalogTable[_, _] =>
         val resolvedTable = lookupResult.getResolvedTable.asInstanceOf[ResolvedCatalogTable]
+        checkManagedTableCheckpointing(null, resolvedTable, objectIdentifier)
```
Comment (JingsongLi) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala:374]: No need here, `ConnectorCatalogTable` is deprecated

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala [hunk 3]
```diff
@@ -533,4 +536,44 @@ abstract class PlannerBase(
     configuration.removeConfig(TABLE_QUERY_START_EPOCH_TIME)
     configuration.removeConfig(TABLE_QUERY_START_LOCAL_TIME)
   }
+
+  private def checkManagedTableCheckpointing(
+                                              catalog: Option[Catalog],
+                                              table: ResolvedCatalogBaseTable[_],
+                                              objectIdentifier: ObjectIdentifier): Unit = {
+    if (isManagedTable(catalog, table) &&
+      !executor.asInstanceOf[DefaultExecutor]
```
Comment (JingsongLi) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala:545]: Can we add a method to `Executor`: `isCheckpointingEnabled`

**Review by JingsongLi — CHANGES_REQUESTED** [review]
@SteNicholas Thanks for your contribution, left some comments, can you add plan test for this?

**Comment by SteNicholas** [comment]
@JingsongLi , thanks for detailed review. I have addressed above comments and added the managed table sink test. Please take a look.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala [hunk 4]
```diff
@@ -384,6 +385,13 @@ abstract class PlannerBase(
         }
         val catalog = toScala(catalogManager.getCatalog(objectIdentifier.getCatalogName))
         val isTemporary = lookupResult.isTemporary
+
+        if (isManagedTable(catalog.get, resolvedTable) &&
+          !executor.isCheckpointingEnabled) {
+          throw new TableException(
+            s"Managed sink ${objectIdentifier} should enable the checkpointing.")
```
Comment (JingsongLi) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala:392]: `You should enable the checkpointing for managed table ${objectIdentifier}, managed table relies on checkpoint to commit and the data is visible only after commit.`
What do you think?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableSinkTest.scala [hunk 5]
```diff
@@ -623,6 +623,32 @@ class TableSinkTest extends TableTestBase {
     // source and sink has same parallelism, but sink shuffle by pk is enforced
     util.verifyExplain(stmtSet, ExplainDetail.JSON_EXECUTION_PLAN)
   }
+
+  @Test
+  def testManagedTableSinkWithDisableCheckpointing(): Unit = {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableSinkTest.scala:628]: Can you add more tests:
- streaming mode with cp
- streaming mode without cp
- batch mode

**Comment by SteNicholas** [comment]
@JingsongLi , thanks for the review. I have addressed your comments and added the tests above mentioned. Please check again.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/TableSinkTest.scala [hunk 6]
```diff
@@ -135,4 +135,45 @@ class TableSinkTest extends TableTestBase {
 
     util.verifyExecPlan(stmtSet)
   }
+
+  @Test
+  def testManagedTableSinkWithDisableCheckpointing(): Unit = {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/TableSinkTest.scala:140]: Batch job can not open checkpoint.
We should not check for batch job.

**Review by JingsongLi — APPROVED** [review]
Thanks for the update. Looks good to me!

**Pull request merged**